pkg/webhook/pod/tikv_util.go

```diff
@@ -0,0 +1,152 @@
+		pod, err := podLister.Pods(namespace).Get(podName)
+		if err != nil {
+			return err
+		}
+		revision, exist := pod.Labels[apps.ControllerRevisionHashLabelKey]
+		if !exist {
+			return fmt.Errorf("tidbcluster: [%s/%s]'s pd pod: [%s] has no label: %s", namespace, tcName, podName, apps.ControllerRevisionHashLabelKey)
+		}
+
+		if revision != tc.Status.TiKV.StatefulSet.UpdateRevision {
+			return fmt.Errorf("tc[%s/%s]'s tikv pod[%s/%s] is not upgraded yet", namespace, tcName, namespace, podName)
+		}
+
+		storeInfo, err := getStoreByPod(pod, tc, storesInfo)
+		if err != nil {
+			return err
+		}
+		klog.Infof("pod[%s/%s] store[%d]'s state is %s", namespace, podName, storeInfo.Store.Id, storeInfo.Store.StateName)
+		if storeInfo.Store.StateName != v1alpha1.TiKVStateUp {
+			return fmt.Errorf("tc[%s/%s]'s tikv pod[%s/%s] state is not up", namespace, tcName, namespace, podName)
+		}
+	}
+	return nil
+}
+
+func addEvictLeaderAnnotation(kubeCli kubernetes.Interface, pod *core.Pod) error {
+
+	name := pod.Name
+	namespace := pod.Namespace
+
+	if pod.Annotations == nil {
+		pod.Annotations = map[string]string{}
+	}
+	now := time.Now().Format(time.RFC3339)
+	pod.Annotations[EvictLeaderBeginTime] = now
+	_, err := kubeCli.CoreV1().Pods(namespace).Update(pod)
+	if err != nil {
+		return err
+	}
+	klog.Infof("tikv upgrader: set pod %s/%s annotation %s to %s successfully",
+		namespace, name, EvictLeaderBeginTime, now)
+
+	return nil
+}
+
+func isTiKVReadyToUpgrade(upgradePod *core.Pod, store *pdapi.StoreInfo) bool {
+
+	if store.Status.LeaderCount == 0 {
+		klog.Infof("pod[%s/%s] has no region leader in store[%d]", upgradePod.Namespace, upgradePod.Name, store.Store.Id)
+		return true
+	}
+
+	if evictLeaderBeginTimeStr, evicting := upgradePod.Annotations[EvictLeaderBeginTime]; evicting {
+		evictLeaderBeginTime, err := time.Parse(time.RFC3339, evictLeaderBeginTimeStr)
+		if err != nil {
+			klog.Errorf("parse annotation:[%s] to time failed.", EvictLeaderBeginTime)
+			return false
+		}
+		if time.Now().After(evictLeaderBeginTime.Add(EvictLeaderTimeout)) {
+			return true
+		}
+	}
+	return false
+}
+
+func beginEvictLeader(kubeCli kubernetes.Interface, storeID uint64, pod *core.Pod, pdClient pdapi.PDClient) error {
+
+	name := pod.Name
+	namespace := pod.Namespace
+
+	err := pdClient.BeginEvictLeader(storeID)
+	if err != nil {
+		klog.Errorf("tikv upgrader: failed to begin evict leader: %d, %s/%s, %v",
+			storeID, namespace, name, err)
+		return err
+	}
+
+	klog.Infof("tikv upgrader: begin evict leader: %d, %s/%s successfully", storeID, namespace, name)
+	err = addEvictLeaderAnnotation(kubeCli, pod)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func endEvictLeader(storeInfo *pdapi.StoreInfo, pdClient pdapi.PDClient) error {
+	storeID := storeInfo.Store.Id
+	err := pdClient.EndEvictLeader(storeID)
+	if err != nil {
+		klog.Errorf("tikv upgrader: failed to end evict leader storeID: %d, %v", storeID, err)
+		return err
```

Comment:
Is there any force strategy for scenarios like this? This pattern introduce a hard dependency from TidbCluster to operator (webhook), if the webhook suffers an network partition to PD (meanwhile, tikv could talk to PD), the Pod creation is blocked by this single point.
As a comparison, former controller is working in bypass and the cluster is still functional without operator.

Reply:
I think `force create` is a nice and necessary choice for this case. Currently `deleting hook pod` is also have a `force` choice. I think it is better to leave it in anthoer issue.